datastore/additional_data/sources/nspl.py

```diff
@@ -0,0 +1,166 @@
+
+    def import_nspl(self, url=NSPL_URL):
+        """
+        Example of a data db entry:
+        {
+            'ccg': 'S03000012', 'ced': None, 'cty': None, 'eer': 'S15000001', 'imd': 6808, 'lat': 57.101474,
+            'pcd': 'AB1 0AA', 'pct': 'S03000012', 'pfa': 'S23000009', 'rgn': None, 'stp': None, 'ctry': 'S92000003',
+            'hash': '3b7db6a231babf449f1c6d7a61ce317f', 'laua': 'S12000033', 'lep1': None, 'lep2': None,
+            'long': -2.242851, 'nuts': 'S31000935', 'oa11': 'S00090303', 'park': None, 'pcd2': 'AB1  0AA',
+            'pcds': 'AB1 0AA', 'pcon': 'S14000002', 'ttwa': 'S22000047', 'ward': 'S13002843', 'wz11': 'S34002990',
+            'bua11': None, 'nhser': None, 'oac11': '1C3', 'calncv': None, 'dointr': '1980-01-01 00:00:00',
+            'doterm': '1996-06-01 00:00:00', 'hlthau': 'S08000020', 'lsoa11': 'S01006514', 'msoa11': 'S02001237',
+            'teclec': 'S09000001', 'buasd11': None, 'ru11ind': '3', 'location': {'lat': 57.101474, 'lon': -2.242851},
+            'oseast1m': 385386, 'osgrdind': 1, 'osnrth1m': 801193, 'usertype': 0
+        }
+        Field names information can be found in NSPL User Guide at https://geoportal.statistics.gov.uk/
+        """
+        zip_file = self.get_zipfile(url)
+        data = self.get_location_data(zip_file)
+        data_json = json.loads(json.dumps(data))
+
+        if NSPL.objects.exists():
+            NSPL.objects.all().delete()
+        self.save_nspl_data(data_json)
+
+    def get_location_data_by_postcode(self, postcode):
+        format_postcode = "".join(postcode.split()).upper()
+        try:
+            nspl = NSPL.objects.get(postcode=format_postcode)
+            return nspl.data
+        except NSPL.DoesNotExist:
+            pass
+
+    def get_location_data_code_names(self, location_data):
+        """Adds new entries to location data dict with code names."""
+        for field_name, field_value in location_data.copy().items():
+            try:
+                code_name_obj = CodeName.objects.get(code=field_value)
+            except CodeName.DoesNotExist:
+                continue
+
+            code_name = code_name_obj.data.get("name")
+            location_data["{}_name".format(field_name)] = code_name
+
+        return location_data
+
+    def update_additional_data(self, grant, additional_data):
+        """Updates with 'recipientOrganizationLocation' based on it's postcode."""
+        recipient_orgs = grant.get("recipientOrganization", [])
+
+        if recipient_orgs:
+            for recipient_org in recipient_orgs:
+                postcode = recipient_org.get("postalCode")
+                if postcode:
+                    location_data = self.get_location_data_by_postcode(postcode)
+                    if location_data:
+                        self.get_location_data_code_names(location_data)
+                        additional_data["recipientOrganizationLocation"] = location_data
```

Comment:
should there be a return after the location data is found to break the loop? Otherwise it will just write the last recipient org it finds (albeit there's usually only one).

Reply:
In Django the query `Model.objects.get(...)` breaks when it finds one. When there is more than one field then it is used `Model.objects.filter(....)`. This last one creates a list of objects.

Reply:
But everytime the loop in line 160 moves to the next iteration, a new value overwrites the old one in `additional_data["recipientOrganizationLocation"]`, rather than adding to it. I don't think we need all of the potential locations to be included, so can't we just return after the first one has been found?

This is especially relevant if we added the lookup to `additional_data["recipientOrgInfos"]` in line 157 above - we want the postcode supplied with the grant to take precedence over a postcode from the organisation info we've brought it.

Reply:
Sorry @drkane , I don't follow what you are saying. Values are not overwritten in this function.

Reply:
Sorry, I haven't been very clear.

The aim of this function is to set `additional_data["recipientOrganizationLocation"]`. It iterates through each instance of `recipientOrganization` in the grant, looks up the location for each one, and sets `additional_data["recipientOrganizationLocation"]` to that location. 

If there are 2 `recipient_orgs` then the `additional_data["recipientOrganizationLocation"]` would first be set to the location of the first one, then overwritten with the location of the second one, and when the function finished then `additional_data["recipientOrganizationLocation"]` would hold the value from the second one (and that's what would end up in the database).

This wouldn't normally come up very often - we've never had a grant with more than 1 `recipientOrganization` attached. However, I propose that at the start of the  function we add the records from `additional_data.get("recipientOrgInfos", [])` to this list. This would mean that if the grant publisher didn't include a postcode, we could instead use the postcode provided by the Charity Commission.

But we want the first entry in `recipient_orgs` to take priority, because that's the one supplied by the grant publisher themselves.

So my proposal is that as soon as `additional_data["recipientOrganizationLocation"]` has been set we return from the function, so that the first value takes precedence. It should also speed things up slightly as location is only checked if a value hasn't already been found.

Reply:
Thanks @drkane ! Now I understand what you mean.

Reply:
I have added the break.